File: go-honeypot/server/internal/honeypots/ftp.go
package honeypots

import (
	"bufio"
	"context"
	"net"
	"strings"

	"github.com/example/honeypot/internal/config"
	"github.com/example/honeypot/internal/eventlog"
)

func NewFTP(cfg config.Service, store *eventlog.Store) *TCP {
	banner := cfg.Banner
	if banner == "" {
		banner = "220 (vsFTPd 3.0.5)\r\n"
	}
	return NewTCP("ftp", cfg, store, func(ctx context.Context, conn net.Conn, sessionID string, meta ConnMeta) {
		_, _ = conn.Write([]byte(banner))
		r := bufio.NewReader(conn)
		var username string
		for {
			line, err := r.ReadString('\n')
			if err != nil {
				return
			}
			line = strings.TrimRight(line, "\r\n")
			cmd, arg := splitFtpLine(line)
			upper := strings.ToUpper(cmd)
			store.Log(eventlog.Event{
				Service: "ftp", Type: "command", SessionID: sessionID,
				RemoteIP: meta.RemoteIP, RemotePort: meta.RemotePort, LocalPort: meta.LocalPort,
				Command: line, Username: username,
				Details: map[string]any{"cmd": upper, "arg": arg},
			})
			switch upper {
			case "USER":
				username = arg
				_, _ = conn.Write([]byte("331 Please specify the password.\r\n"))
			case "PASS":
				store.Log(eventlog.Event{
					Service: "ftp", Type: "login_attempt", SessionID: sessionID,
					RemoteIP: meta.RemoteIP, RemotePort: meta.RemotePort, LocalPort: meta.LocalPort,
					Username: username, Password: arg,
				})
				_, _ = conn.Write([]byte("530 Login incorrect.\r\n"))
			case "SYST":
				_, _ = conn.Write([]byte("215 UNIX Type: L8\r\n"))
			case "FEAT":
				_, _ = conn.Write([]byte("211-Features:\r\n PASV\r\n UTF8\r\n211 End\r\n"))
			case "AUTH":
				_, _ = conn.Write([]byte("530 Please login with USER and PASS.\r\n"))
			case "QUIT":
				_, _ = conn.Write([]byte("221 Goodbye.\r\n"))
				return
			default:
				_, _ = conn.Write([]byte("530 Please login with USER and PASS.\r\n"))
			}
			_ = ctx
		}
	})
}

func splitFtpLine(line string) (string, string) {
	idx := strings.Index(line, " ")
	if idx < 0 {
		return line, ""
	}
	return line[:idx], line[idx+1:]
}

File: src/honeypots/ftp.js
'use strict';

const { TcpHoneypot } = require('./base');

class FtpHoneypot extends TcpHoneypot {
  constructor(cfg, logger) {
    super({ name: 'ftp', port: cfg.port }, logger);
    this.cfg = cfg;
  }

  updateConfig(cfg) { this.cfg = cfg; }

  onConnection(socket, meta) {
    let username = null;
    let buf = '';
    const send = (s) => { try { socket.write(s); } catch { /* noop */ } };

    send(this.cfg.banner || '220 (vsFTPd 3.0.5)\r\n');

    socket.on('data', (data) => {
      buf += data.toString('utf8');
      let idx;
      while ((idx = buf.indexOf('\n')) >= 0) {
        const line = buf.slice(0, idx).replace(/\r$/, '');
        buf = buf.slice(idx + 1);
        this._handleLine(socket, meta, line, (u) => { username = u; }, () => username);
      }
    });
  }

  _handleLine(socket, meta, line, setUser, getUser) {
    const send = (s) => { try { socket.write(s + '\r\n'); } catch { /* noop */ } };
    const spaceIdx = line.indexOf(' ');
    const cmd = (spaceIdx >= 0 ? line.slice(0, spaceIdx) : line).toUpperCase();
    const arg = spaceIdx >= 0 ? line.slice(spaceIdx + 1) : '';

    this.logger.log({
      service: 'ftp',
      type: 'command',
      sessionId: meta.sessionId,
      remoteIp: meta.remoteIp,
      remotePort: meta.remotePort,
      localPort: meta.localPort,
      command: line,
      username: getUser(),
      details: { cmd, arg }
    });

    switch (cmd) {
      case 'USER':
        setUser(arg);
        send('331 Please specify the password.');
        break;
      case 'PASS':
        this.logger.log({
          service: 'ftp',
          type: 'login_attempt',
          sessionId: meta.sessionId,
          remoteIp: meta.remoteIp,
          remotePort: meta.remotePort,
          localPort: meta.localPort,
          username: getUser(),
          password: arg
        });
        send('530 Login incorrect.');
        break;
      case 'SYST':
        send('215 UNIX Type: L8');
        break;
      case 'FEAT':
        send('211-Features:\r\n PASV\r\n UTF8\r\n211 End');
        break;
      case 'AUTH':
        send('530 Please login with USER and PASS.');
        break;
      case 'QUIT':
        send('221 Goodbye.');
        try { socket.end(); } catch { /* noop */ }
        break;
      default:
        send('530 Please login with USER and PASS.');
    }
  }
}

module.exports = { create: (cfg, logger) => new FtpHoneypot(cfg, logger) };
